### ISSUE TITLE
Java Grammar: Hard to understand message

### ISSUE THREAD

**Issue by denizarsan** [checkstyle/checkstyle#4632]
$ cat C.java # the code does not compile
```java
class C extends D<${package}.E> {}
```

$ cat config.xml
```
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="LineLength"/> <!-- can be any check that requires parsing -->
  </module>
</module>

```
$ java -jar checkstyle-8.0-all.jar -c config.xml C.java
```
Starting audit...
 areLtsAndGtsBalanced(currentLtLevel)
Audit done.
```
---------------
The source file does not compile, so giving a warning/error message is fine, but the message is hard to understand.

**Comment by rnveach** [comment]
I can reproduce the message and it is not very informative. It doesn't even look like an error, but is most likely a parsing error as file is uncompilable.
We should display some type of error about failing to parse like javadoc.
@romani You agree?

**Comment by romani** [comment]
We can not detect non compiled sources at all. We could even parse without problems some of them, we could parse them but Checks could fail.
There are a lot of uncertainty could happen if source is not compilable. 
That is why it is our limitation to work only on compiled sources.

I am closing issue , but discussion could continue

**Comment by rnveach** [comment]
@romani The main issue is that CS isn't throwing this as an error or violation. Except for that one display, TreeWalker assumes the file was parsed correctly even though the AST being returned is `null`. We have places with null protection on the tree being returned to not throw an NPE.
If cache was enabled, you would only see the output once as it would add the file to the cache since there is no violation.

This is the exception being produced in master:
````
 areLtsAndGtsBalanced(currentLtLevel)
	at com.puppycrawl.tools.checkstyle.grammars.GeneratedJavaRecognizer.typeArguments(GeneratedJavaRecognizer.java:1086)
	at com.puppycrawl.tools.checkstyle.grammars.GeneratedJavaRecognizer.classOrInterfaceType(GeneratedJavaRecognizer.java:958)
	at com.puppycrawl.tools.checkstyle.grammars.GeneratedJavaRecognizer.superClassClause(GeneratedJavaRecognizer.java:2425)
	at com.puppycrawl.tools.checkstyle.grammars.GeneratedJavaRecognizer.classDefinition(GeneratedJavaRecognizer.java:630)
	at com.puppycrawl.tools.checkstyle.grammars.GeneratedJavaRecognizer.typeDefinitionInternal(GeneratedJavaRecognizer.java:556)
	at com.puppycrawl.tools.checkstyle.grammars.GeneratedJavaRecognizer.typeDefinition(GeneratedJavaRecognizer.java:389)
	at com.puppycrawl.tools.checkstyle.grammars.GeneratedJavaRecognizer.compilationUnit(GeneratedJavaRecognizer.java:202)
	at com.puppycrawl.tools.checkstyle.TreeWalker.parse(TreeWalker.java:445)
````
However, `typeDefinition` catches this exception and only does a simple print to `System.err` via `reportError(ex)` and than continues on by returning `null`.
We should look into not hiding the failure.

**Comment by romani** [comment]
yes, lets review this case, but expected behavior is any wide exception or message, it still could be unclear to user, but we should not hide details.

**Comment by rnveach** [comment]
Fix was merged
